docs/proposals/connector-protocol-v2.md

```diff
@@ -325,13 +324,9 @@ This is a newer feature and it aligns nicely with GenSource. Generators and batc
 
 The worker should acknowledge frames of all types. These do not need to be 1-1 acknowledgments but it may be convenient to write it as such initially since all frames take one credit and each session needs to have these continuously replenished to avoid stuttered flow <sup>D2</sup>. Ack frames serve this purpose, almost like a regular heartbeat. Technically an ack could be sent with zero credits but we don't currently see this as a very useful feature <sup>D3</sup>.
 
-Acks also provide a list of message ids for each stream which have completed some safe level of processing in Wallaroo <sup>D4</sup>.
-
-NOTE: All messages ids must be sent back as ack'ed but it may be enough to leverage transmission ordering and ack only the last transmitted in the set for each stream. This may be a future optimization to consider.
+When Wallaroo has processed a barrier and completed a checkpoint, then, for each stream, the last received MessageId prior to the barrier will be sent to the connector client in an ACK message to report processing progress <sup>D4</sup>.  All messages with MessageIds less than the reported point of reference are included in the checkpoint.
 
-Nacks are slightly different from an ack. They provide credits to allow new frames to be transmitted but not represent forward progress. This happens when Wallaroo needs to reset a given stream's state to a specific point of reference after the session has already been started. The stream is put into a nack'ed state in the worker which effectively causes all message frames to be ignored for that stream until a new notify message is sent for that stream to reset it to an open state with a specific point of reference to continue from.
-
-If the connector is unable to resume from the given point of reference then it should not notify but instead send an error and exit the session <sup>D5</sup>.
+When a connector receives a NACK message from the worker, the connector must close the connection.  In order to resume sending data to the worker, the connector must open a new connection + handshake + notify for each stream.  The worker cannot assume that any messages sent with MessageIds larger than the last ACK's point of reference were received by the worker.
```

Comment:
As per my earlier comment, I think an error frame might be better here if we're not going for partial recovery (per-stream rather than per-session).